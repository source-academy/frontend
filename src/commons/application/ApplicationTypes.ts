import { Chapter, Language, SourceError, Variant } from 'js-slang/dist/types';

import { AcademyState } from '../../features/academy/AcademyTypes';
import { AchievementState } from '../../features/achievement/AchievementTypes';
import { DashboardState } from '../../features/dashboard/DashboardTypes';
import { Grading } from '../../features/grading/GradingTypes';
import { PlaygroundState } from '../../features/playground/PlaygroundTypes';
import { PlaybackStatus, RecordingStatus } from '../../features/sourceRecorder/SourceRecorderTypes';
import { WORKSPACE_BASE_PATHS } from '../../pages/fileSystem/createInBrowserFileSystem';
import { Assessment } from '../assessment/AssessmentTypes';
import { FileSystemState } from '../fileSystem/FileSystemTypes';
import Constants from '../utils/Constants';
import { createContext } from '../utils/JsSlangHelper';
import {
  DebuggerContext,
  WorkspaceLocation,
  WorkspaceManagerState,
  WorkspaceState
} from '../workspace/WorkspaceTypes';
import { ExternalLibraryName } from './types/ExternalTypes';
import { SessionState } from './types/SessionTypes';

export type OverallState = {
  readonly academy: AcademyState;
  readonly achievement: AchievementState;
  readonly application: ApplicationState;
  readonly playground: PlaygroundState;
  readonly session: SessionState;
  readonly workspaces: WorkspaceManagerState;
  readonly dashboard: DashboardState;
  readonly fileSystem: FileSystemState;
};

export type ApplicationState = {
  readonly environment: ApplicationEnvironment;
};

export type Story = {
  story: string;
  playStory: boolean;
};

export type GameState = {
  collectibles: { [id: string]: string };
  completed_quests: string[];
};

/**
 * An output while the program is still being run in the interpreter. As a
 * result, there are no return values or SourceErrors yet. However, there could
 * have been calls to display (console.log) that need to be printed out.
 */
export type RunningOutput = {
  type: 'running';
  consoleLogs: string[];
};

/**
 * An output which reflects the program which the user had entered. Not a true
 * Output from the interpreter, but simply there to let he user know what had
 * been entered.
 */
export type CodeOutput = {
  type: 'code';
  value: string;
};

/**
 * An output which represents a program being run successfully, i.e. with a
 * return value at the end. A program can have either a return value, or errors,
 * but not both.
 */
export type ResultOutput = {
  type: 'result';
  value: any;
  consoleLogs: string[];
  runtime?: number;
  isProgram?: boolean;
};

/**
 * An output which represents a program being run unsuccessfully, i.e. with
 * errors at the end. A program can have either a return value, or errors, but
 * not both.
 */
export type ErrorOutput = {
  type: 'errors';
  errors: SourceError[];
  consoleLogs: string[];
};

export type InterpreterOutput = RunningOutput | CodeOutput | ResultOutput | ErrorOutput;

export enum ApplicationEnvironment {
  Development = 'development',
  Production = 'production',
  Test = 'test'
}

export enum Role {
  Student = 'student',
  Staff = 'staff',
  Admin = 'admin'
}

/**
 * Defines the languages available for use on Source Academy,
 * including Source sublanguages and other languages e.g. full JS.
 * For external libraries, see ExternalLibrariesTypes.ts
 */
export interface SALanguage extends Language {
  displayName: string;
}

const variantDisplay: Map<Variant, string> = new Map([
  [Variant.TYPED, 'Typed'],
  [Variant.WASM, 'WebAssembly'],
  [Variant.NON_DET, 'Non-Det'],
  [Variant.CONCURRENT, 'Concurrent'],
  [Variant.LAZY, 'Lazy'],
  [Variant.GPU, 'GPU'],
  [Variant.NATIVE, 'Native'],
  [Variant.EXPLICIT_CONTROL, 'Explicit-Control']
]);

export const fullJSLanguage: SALanguage = {
  chapter: Chapter.FULL_JS,
  variant: Variant.DEFAULT,
  displayName: 'full JavaScript'
};

export const fullTSLanguage: SALanguage = {
  chapter: Chapter.FULL_TS,
  variant: Variant.DEFAULT,
  displayName: 'full TypeScript'
};

export const htmlLanguage: SALanguage = {
  chapter: Chapter.HTML,
  variant: Variant.DEFAULT,
  displayName: 'HTML'
};

export const styliseSublanguage = (chapter: Chapter, variant: Variant = Variant.DEFAULT) => {
  switch (chapter) {
    case Chapter.FULL_JS:
      return fullJSLanguage.displayName;
    case Chapter.FULL_TS:
      return fullTSLanguage.displayName;
    case Chapter.HTML:
      return htmlLanguage.displayName;
    default:
      return `Source \xa7${chapter}${
        variantDisplay.has(variant) ? ` ${variantDisplay.get(variant)}` : ''
      }`;
  }
};

export const sublanguages: Language[] = [
  { chapter: Chapter.SOURCE_1, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_1, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_1, variant: Variant.WASM },
  { chapter: Chapter.SOURCE_1, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_1, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_2, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_2, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_2, variant: Variant.LAZY },
  { chapter: Chapter.SOURCE_2, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_3, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_3, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_3, variant: Variant.CONCURRENT },
  { chapter: Chapter.SOURCE_3, variant: Variant.NON_DET },
  { chapter: Chapter.SOURCE_3, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.DEFAULT },
  { chapter: Chapter.SOURCE_4, variant: Variant.TYPED },
  { chapter: Chapter.SOURCE_4, variant: Variant.GPU },
  { chapter: Chapter.SOURCE_4, variant: Variant.NATIVE },
  { chapter: Chapter.SOURCE_4, variant: Variant.EXPLICIT_CONTROL }
];

export const sourceLanguages: SALanguage[] = sublanguages.map(sublang => {
  return {
    ...sublang,
    displayName: styliseSublanguage(sublang.chapter, sublang.variant)
  };
});

export const defaultLanguages = sourceLanguages.filter(
  sublang => sublang.variant === Variant.DEFAULT
);

export const variantLanguages = sourceLanguages.filter(
  sublang => sublang.variant !== Variant.DEFAULT
);

export const isSourceLanguage = (chapter: Chapter) =>
  [Chapter.SOURCE_1, Chapter.SOURCE_2, Chapter.SOURCE_3, Chapter.SOURCE_4].includes(chapter);

const currentEnvironment = (): ApplicationEnvironment => {
  switch (process.env.NODE_ENV) {
    case 'development':
      return ApplicationEnvironment.Development;
    case 'production':
      return ApplicationEnvironment.Production;
    default:
      return ApplicationEnvironment.Test;
  }
};

export const defaultAcademy: AcademyState = {
  gameCanvas: undefined
};

export const defaultApplication: ApplicationState = {
  environment: currentEnvironment()
};

export const defaultDashboard: DashboardState = {
  gradingSummary: {
    cols: [],
    rows: []
  }
};

export const defaultAchievement: AchievementState = {
  achievements: [],
  goals: [],
  users: [],
  assessmentOverviews: []
};

export const defaultPlayground: PlaygroundState = {
  githubSaveInfo: { repoName: '', filePath: '' }
};

export const defaultEditorValue = '// Type your program in here!';

/**
 * Create a default IWorkspaceState for 'resetting' a workspace.
 * Takes in parameters to set the js-slang library and chapter.
 *
 * @param workspaceLocation the location of the workspace, used for context
 */
export const createDefaultWorkspace = (workspaceLocation: WorkspaceLocation): WorkspaceState => ({
  autogradingResults: [],
  context: createContext<WorkspaceLocation>(
    Constants.defaultSourceChapter,
    [],
    workspaceLocation,
    Constants.defaultSourceVariant
  ),
  isFolderModeEnabled: false,
  activeEditorTabIndex: 0,
  editorTabs: [
    {
      filePath: ['playground', 'sicp'].includes(workspaceLocation)
        ? getDefaultFilePath(workspaceLocation)
        : undefined,
      value: ['playground', 'sourcecast', 'githubAssessments'].includes(workspaceLocation)
        ? defaultEditorValue
        : '',
      highlightedLines: [],
      breakpoints: []
    }
  ],
  programPrependValue: '',
  programPostpendValue: '',
  editorSessionId: '',
  isEditorReadonly: false,
  editorTestcases: [],
  externalLibrary: ExternalLibraryName.NONE,
  execTime: 1000,
  output: [],
  replHistory: {
    browseIndex: null,
    records: [],
    originalValue: ''
  },
  replValue: '',
  sharedbConnected: false,
  stepLimit: 1000,
  globals: [],
  isEditorAutorun: false,
  isRunning: false,
  isDebugging: false,
  enableDebugging: true,
  debuggerContext: {} as DebuggerContext
});

const defaultFileName = 'program.js';
export const getDefaultFilePath = (workspaceLocation: WorkspaceLocation) =>
  `${WORKSPACE_BASE_PATHS[workspaceLocation]}/${defaultFileName}`;

export const defaultWorkspaceManager: WorkspaceManagerState = {
  assessment: {
    ...createDefaultWorkspace('assessment'),
    currentAssessment: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  grading: {
    ...createDefaultWorkspace('grading'),
    currentSubmission: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  playground: {
    ...createDefaultWorkspace('playground'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('playground'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  sourcecast: {
    ...createDefaultWorkspace('sourcecast'),
    audioUrl: '',
    codeDeltasToApply: null,
    currentPlayerTime: 0,
    description: null,
    inputToApply: null,
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    playbackDuration: 0,
    playbackStatus: PlaybackStatus.paused,
    sourcecastIndex: null,
    title: null,
    uid: null
  },
  sourcereel: {
    ...createDefaultWorkspace('sourcereel'),
    playbackData: {
      init: {
        editorValue: '',
        chapter: Chapter.SOURCE_1,
        externalLibrary: ExternalLibraryName.NONE
      },
      inputs: []
    },
    recordingStatus: RecordingStatus.notStarted,
    timeElapsedBeforePause: 0,
    timeResumed: 0
  },
  sicp: {
    ...createDefaultWorkspace('sicp'),
    usingSubst: false,
    activeEditorTabIndex: 0,
    editorTabs: [
      {
        filePath: getDefaultFilePath('sicp'),
        value: defaultEditorValue,
        highlightedLines: [],
        breakpoints: []
      }
    ]
  },
  githubAssessment: {
    ...createDefaultWorkspace('githubAssessment'),
    hasUnsavedChanges: false
  }
};

export const defaultSession: SessionState = {
  courses: [],
  group: null,
  gameState: {
    completed_quests: [],
    collectibles: {}
  },
  xp: 0,
  allUserXp: undefined,
  story: {
    story: '',
    playStory: false
  },
  assessments: new Map<number, Assessment>(),
  assessmentOverviews: undefined,
  agreedToResearch: undefined,
  sessionId: Date.now(),
  githubOctokitObject: { octokit: undefined },
  gradingOverviews: undefined,
  gradings: new Map<number, Grading>(),
  notifications: []
};

export const defaultFileSystem: FileSystemState = {
  inBrowserFileSystem: null
};

export const defaultState: OverallState = {
  academy: defaultAcademy,
  achievement: defaultAchievement,
  application: defaultApplication,
  dashboard: defaultDashboard,
  playground: defaultPlayground,
  session: defaultSession,
  workspaces: defaultWorkspaceManager,
  fileSystem: defaultFileSystem
};
