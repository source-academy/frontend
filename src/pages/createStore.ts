import { throttle } from 'lodash';
import { applyMiddleware, compose, createStore as _createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import { defaultState } from '../commons/application/ApplicationTypes';
import createRootReducer from '../commons/application/reducers/RootReducer';
import MainSaga from '../commons/sagas/MainSaga';
import { generateOctokitInstance } from '../commons/utils/GitHubPersistenceHelper';
import { loadStoredState, SavedState, saveState } from './localStorage';

export const store = createStore();

export function createStore() {
  const sagaMiddleware = createSagaMiddleware();
  const middleware = [sagaMiddleware];

  const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
        serialize: true,
        maxAge: 300
      }) || compose
    : compose;

  const initialStore = loadStore(loadStoredState()) || defaultState;

  const enhancers = composeEnhancers(applyMiddleware(...middleware));

  const createdStore = _createStore(createRootReducer(), initialStore as any, enhancers);
  sagaMiddleware.run(MainSaga);

  createdStore.subscribe(
    throttle(() => {
      saveState(createdStore.getState());
    }, 1000)
  );

  return createdStore;
}

function loadStore(loadedStore: SavedState | undefined) {
  if (!loadedStore) {
    return undefined;
  }
  return {
    ...defaultState,
    session: {
      ...defaultState.session,
      ...(loadedStore.session ? loadedStore.session : {}),
      githubOctokitObject: {
        octokit: loadedStore.session.githubAccessToken
          ? generateOctokitInstance(loadedStore.session.githubAccessToken)
          : undefined
      }
    },
    workspaces: {
      ...defaultState.workspaces,
      playground: {
        ...defaultState.workspaces.playground,
        isFolderModeEnabled: loadedStore.playgroundIsFolderModeEnabled
          ? loadedStore.playgroundIsFolderModeEnabled
          : defaultState.workspaces.playground.isFolderModeEnabled,
        activeEditorTabIndex: loadedStore.playgroundActiveEditorTabIndex
          ? loadedStore.playgroundActiveEditorTabIndex.value
          : defaultState.workspaces.playground.activeEditorTabIndex,
        editorTabs: loadedStore.playgroundEditorTabs
          ? loadedStore.playgroundEditorTabs
          : defaultState.workspaces.playground.editorTabs,
        isEditorAutorun: loadedStore.playgroundIsEditorAutorun
          ? loadedStore.playgroundIsEditorAutorun
          : defaultState.workspaces.playground.isEditorAutorun,
        externalLibrary: loadedStore.playgroundExternalLibrary
          ? loadedStore.playgroundExternalLibrary
          : defaultState.workspaces.playground.externalLibrary,
        context: {
          ...defaultState.workspaces.playground.context,
          chapter: loadedStore.playgroundSourceChapter
            ? loadedStore.playgroundSourceChapter
            : defaultState.workspaces.playground.context.chapter,
          variant: loadedStore.playgroundSourceVariant
            ? loadedStore.playgroundSourceVariant
            : defaultState.workspaces.playground.context.variant
        }
      }
    }
  };
}
