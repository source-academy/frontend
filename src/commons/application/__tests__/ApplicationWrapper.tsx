import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMemoryRouter, RouterProvider } from 'react-router';
import { createStore } from 'src/pages/createStore';
import { getFullAcademyRouterConfig, playgroundOnlyRouterConfig } from 'src/routes/routerConfig';

import { Role } from '../ApplicationTypes';

// JSDOM does not implement window.matchMedia, so we have to mock it.
window.matchMedia =
  window.matchMedia ||
  function () {
    return {
      matches: false,
      addListener: function () {},
      removeListener: function () {}
    };
  };

describe('ApplicationWrapper', () => {
  let store = createStore();
  beforeEach(() => {
    store = createStore();
  });

  test('ApplicationWrapper renders NotFound on unknown routes (Full Academy)', async () => {
    const routerConfig = getFullAcademyRouterConfig({
      name: 'Bob',
      role: Role.Student,
      isLoggedIn: false,
      courseId: 1
    });

    const app = (
      <Provider store={store}>
        <RouterProvider
          router={createMemoryRouter(routerConfig, { initialEntries: ['/unknown'] })}
        />
      </Provider>
    );
    render(app);
    const element = await waitFor(() => screen.getByTestId('NotFound-Component'));
    expect(element).toBeTruthy();
  });

  test('ApplicationWrapper renders NotFound on unknown routes (Playground Only)', async () => {
    const routerConfig = playgroundOnlyRouterConfig;

    const app = (
      <Provider store={store}>
        <RouterProvider
          router={createMemoryRouter(routerConfig, { initialEntries: ['/unknown'] })}
        />
      </Provider>
    );
    render(app);
    const element = await waitFor(() => screen.getByTestId('NotFound-Component'));
    expect(element).toBeTruthy();
  });
});
