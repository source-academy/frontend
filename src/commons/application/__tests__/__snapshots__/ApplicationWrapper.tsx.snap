// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationWrapper Application shows disabled when in disabled period 1`] = `
"<Provider store={{...}}>
  <ApplicationWrapper>
    <RouterProvider router={{...}}>
      <Router basename=\\"/\\" location={{...}} navigationType=\\"POP\\" navigator={{...}}>
        <DataRoutes routes={{...}} state={{...}}>
          <RenderErrorBoundary location={{...}} revalidation=\\"idle\\" component={{...}} error={[undefined]} routeContext={{...}}>
            <RenderedRoute match={{...}} routeContext={{...}}>
              <Application>
                <div className=\\"Application\\">
                  <NavigationBar>
                    <Blueprint4.Navbar className=\\"NavigationBar primary-navbar bp4-dark\\">
                      <div className=\\"bp4-navbar NavigationBar primary-navbar bp4-dark\\">
                        <Blueprint4.NavbarGroup align=\\"left\\">
                          <div className=\\"bp4-navbar-group bp4-align-left\\">
                            <NavLink to=\\"/playground\\" className={[Function: className]}>
                              <Link aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} to=\\"/playground\\">
                                <a aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} href=\\"/playground\\" onClick={[Function: handleClick]} target={[undefined]}>
                                  <Blueprint4.Icon icon=\\"code\\">
                                    <span aria-hidden={true} icon=\\"code\\" className=\\"bp4-icon bp4-icon-code\\" title={[undefined]}>
                                      <svg fill={[undefined]} data-icon=\\"code\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                        <path d=\\"M15.71 7.29l-3-3a1.003 1.003 0 00-1.42 1.42L13.59 8l-2.29 2.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l3-3c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71zM5 5a1.003 1.003 0 00-1.71-.71l-3 3C.11 7.47 0 7.72 0 8c0 .28.11.53.29.71l3 3a1.003 1.003 0 001.42-1.42L2.41 8 4.7 5.71c.19-.18.3-.43.3-.71zm4-3c-.48 0-.87.35-.96.81l-2 10c-.01.06-.04.12-.04.19 0 .55.45 1 1 1 .48 0 .87-.35.96-.81l2-10c.01-.06.04-.12.04-.19 0-.55-.45-1-1-1z\\" fillRule=\\"evenodd\\" />
                                      </svg>
                                    </span>
                                  </Blueprint4.Icon>
                                  <div>
                                    Playground
                                  </div>
                                </a>
                              </Link>
                            </NavLink>
                            <NavLink to=\\"/githubassessments\\" className={[Function: className]}>
                              <Link aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} to=\\"/githubassessments\\">
                                <a aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} href=\\"/githubassessments\\" onClick={[Function: handleClick]} target={[undefined]}>
                                  <Blueprint4.Icon icon=\\"briefcase\\">
                                    <span aria-hidden={true} icon=\\"briefcase\\" className=\\"bp4-icon bp4-icon-briefcase\\" title={[undefined]}>
                                      <svg fill={[undefined]} data-icon=\\"briefcase\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                        <path d=\\"M15 3.98h-3v-2c0-.55-.45-1-1-1H5c-.55 0-1 .45-1 1v2H1c-.55 0-1 .45-1 1v4h3v-1h2v1h6v-1h2v1h3v-4c0-.55-.45-1-1-1zm-5 0H6v-1h4v1zm3 7h-2v-1H5v1H3v-1H0v4c0 .55.45 1 1 1h14c.55 0 1-.45 1-1v-4h-3v1z\\" fillRule=\\"evenodd\\" />
                                      </svg>
                                    </span>
                                  </Blueprint4.Icon>
                                  <div>
                                    Classroom
                                  </div>
                                </a>
                              </Link>
                            </NavLink>
                            <NavLink to=\\"/sicpjs/\\" className={[Function: className]}>
                              <Link aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} to=\\"/sicpjs/\\">
                                <a aria-current={[undefined]} className=\\"NavigationBar__link__mobile bp4-button bp4-minimal\\" style={[undefined]} href=\\"/sicpjs/\\" onClick={[Function: handleClick]} target={[undefined]}>
                                  <Blueprint4.Icon icon=\\"book\\">
                                    <span aria-hidden={true} icon=\\"book\\" className=\\"bp4-icon bp4-icon-book\\" title={[undefined]}>
                                      <svg fill={[undefined]} data-icon=\\"book\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                        <path d=\\"M2 1v14c0 .55.45 1 1 1h1V0H3c-.55 0-1 .45-1 1zm11-1h-1v7l-2-2-2 2V0H5v16h8c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1z\\" fillRule=\\"evenodd\\" />
                                      </svg>
                                    </span>
                                  </Blueprint4.Icon>
                                  <div>
                                    SICP JS
                                  </div>
                                </a>
                              </Link>
                            </NavLink>
                          </div>
                        </Blueprint4.NavbarGroup>
                        <Blueprint4.NavbarGroup align=\\"right\\">
                          <div className=\\"bp4-navbar-group bp4-align-right\\">
                            <NavLink className={[Function: className]} to=\\"/contributors\\">
                              <Link aria-current={[undefined]} className=\\"NavigationBar__link bp4-button bp4-minimal\\" style={[undefined]} to=\\"/contributors\\">
                                <a aria-current={[undefined]} className=\\"NavigationBar__link bp4-button bp4-minimal\\" style={[undefined]} href=\\"/contributors\\" onClick={[Function: handleClick]} target={[undefined]}>
                                  <Blueprint4.Icon icon=\\"heart\\">
                                    <span aria-hidden={true} icon=\\"heart\\" className=\\"bp4-icon bp4-icon-heart\\" title={[undefined]}>
                                      <svg fill={[undefined]} data-icon=\\"heart\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                        <path d=\\"M16 5.095c0-2.255-1.88-4.083-4.2-4.083-1.682 0-3.13.964-3.8 2.352a4.206 4.206 0 00-3.8-2.352C1.88 1.012 0 2.84 0 5.095c0 .066.007.13.01.194H.004c.001.047.01.096.014.143l.013.142c.07.8.321 1.663.824 2.573C2.073 10.354 4.232 12.018 8 15c3.767-2.982 5.926-4.647 7.144-6.854.501-.905.752-1.766.823-2.562.007-.055.012-.11.016-.164.003-.043.012-.088.013-.13h-.006c.003-.066.01-.13.01-.195z\\" fillRule=\\"evenodd\\" />
                                      </svg>
                                    </span>
                                  </Blueprint4.Icon>
                                </a>
                              </Link>
                            </NavLink>
                            <div className=\\"visible-xs\\">
                              <Blueprint4.NavbarDivider className=\\"thin-divider\\">
                                <div className=\\"bp4-navbar-divider thin-divider\\" />
                              </Blueprint4.NavbarDivider>
                            </div>
                            <div className=\\"hidden-xs\\">
                              <Blueprint4.NavbarDivider className=\\"default-divider\\">
                                <div className=\\"bp4-navbar-divider default-divider\\" />
                              </Blueprint4.NavbarDivider>
                            </div>
                            <Dropdown>
                              <Blueprint4.Popover2 content={{...}} inheritDarkTheme={false} placement=\\"bottom\\" boundary=\\"clippingParents\\" captureDismiss={false} defaultIsOpen={false} disabled={false} fill={false} hasBackdrop={false} hoverCloseDelay={300} hoverOpenDelay={150} interactionKind=\\"click\\" matchTargetWidth={false} minimal={false} openOnTargetFocus={true} positioningStrategy=\\"absolute\\" renderTarget={[undefined]} shouldReturnFocusOnClose={false} targetTagName=\\"span\\" transitionDuration={300} usePortal={true}>
                                <Manager>
                                  <Reference>
                                    <Blueprint4.ResizeSensor2 targetRef={[Function (anonymous)]} onResize={[Function (anonymous)]}>
                                      <span aria-haspopup=\\"true\\" className=\\"bp4-popover2-target\\" onClick={[Function (anonymous)]} onKeyDown={[Function: onKeyDown]}>
                                        <ControlButton icon=\\"caret-down\\" className=\\"\\" disabled={[undefined]} tabIndex={[undefined]}>
                                          <Blueprint4.Button disabled={false} fill={false} intent=\\"none\\" minimal={true} className=\\"\\" type={[undefined]} onClick={[undefined]} icon={{...}} rightIcon={false}>
                                            <button type={[undefined]} disabled={false} className=\\"bp4-button bp4-minimal\\" onClick={[undefined]} onBlur={[Function (anonymous)]} onFocus={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                              <Blueprint4.Icon icon={{...}}>
                                                <Blueprint4.Icon icon=\\"caret-down\\" color={[undefined]}>
                                                  <span aria-hidden={true} icon=\\"caret-down\\" className=\\"bp4-icon bp4-icon-caret-down\\" title={[undefined]}>
                                                    <svg fill={[undefined]} data-icon=\\"caret-down\\" width={16} height={16} viewBox=\\"0 0 16 16\\" aria-labelledby={[undefined]} role=\\"img\\">
                                                      <path d=\\"M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z\\" fillRule=\\"evenodd\\" />
                                                    </svg>
                                                  </span>
                                                </Blueprint4.Icon>
                                              </Blueprint4.Icon>
                                              <Blueprint4.Icon icon={false} />
                                            </button>
                                          </Blueprint4.Button>
                                        </ControlButton>
                                      </span>
                                    </Blueprint4.ResizeSensor2>
                                  </Reference>
                                  <Popper innerRef={[Function (anonymous)]} placement=\\"bottom\\" strategy=\\"absolute\\" modifiers={{...}}>
                                    <Blueprint4.Overlay autoFocus={true} backdropClassName=\\"bp4-popover2-backdrop\\" backdropProps={{...}} canEscapeKeyClose={true} canOutsideClickClose={true} enforceFocus={true} hasBackdrop={false} isOpen={false} onClose={[Function (anonymous)]} onClosed={[undefined]} onClosing={[undefined]} onOpened={[undefined]} onOpening={[undefined]} transitionDuration={300} transitionName=\\"bp4-popover2\\" usePortal={true} portalClassName={[undefined]} portalContainer={[undefined]} shouldReturnFocusOnClose={false} lazy={true} />
                                  </Popper>
                                </Manager>
                              </Blueprint4.Popover2>
                              <DropdownSettings isOpen={false} onClose={[Function: toggleSettingsOpen]}>
                                <Blueprint4.Dialog className=\\"settings\\" icon=\\"cog\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleSettingsOpen]} title=\\"Settings\\" canOutsideClickClose={true}>
                                  <Blueprint4.Overlay className=\\"bp4-overlay-scroll-container\\" icon=\\"cog\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleSettingsOpen]} title=\\"Settings\\" canOutsideClickClose={true} hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} shouldReturnFocusOnClose={true} transitionDuration={300} transitionName=\\"bp4-overlay\\" usePortal={true} />
                                </Blueprint4.Dialog>
                              </DropdownSettings>
                              <DropdownAbout isOpen={false} onClose={[Function: toggleAboutOpen]}>
                                <Blueprint4.Dialog className=\\"about\\" icon=\\"help\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleAboutOpen]} title=\\"About\\" canOutsideClickClose={true}>
                                  <Blueprint4.Overlay className=\\"bp4-overlay-scroll-container\\" icon=\\"help\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleAboutOpen]} title=\\"About\\" canOutsideClickClose={true} hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} shouldReturnFocusOnClose={true} transitionDuration={300} transitionName=\\"bp4-overlay\\" usePortal={true} />
                                </Blueprint4.Dialog>
                              </DropdownAbout>
                              <DropdownHelp isOpen={false} onClose={[Function: toggleHelpOpen]}>
                                <Blueprint4.Dialog className=\\"help\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleHelpOpen]} title=\\"Help\\" canOutsideClickClose={true}>
                                  <Blueprint4.Overlay className=\\"bp4-overlay-scroll-container\\" icon=\\"error\\" isCloseButtonShown={true} isOpen={false} onClose={[Function: toggleHelpOpen]} title=\\"Help\\" canOutsideClickClose={true} hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} shouldReturnFocusOnClose={true} transitionDuration={300} transitionName=\\"bp4-overlay\\" usePortal={true} />
                                </Blueprint4.Dialog>
                              </DropdownHelp>
                            </Dropdown>
                          </div>
                        </Blueprint4.NavbarGroup>
                      </div>
                    </Blueprint4.Navbar>
                    <Routes>
                      <RenderedRoute match={{...}} routeContext={{...}} />
                    </Routes>
                  </NavigationBar>
                  <div className=\\"Application__main\\">
                    <Outlet>
                      <RenderedRoute match={{...}} routeContext={{...}}>
                        <Disabled reason=\\"Testing\\">
                          <div className=\\"NoPage bp4-dark\\">
                            <Blueprint4.NonIdealState icon=\\"error\\" title=\\"Disabled\\" description=\\"The Source Academy has been disabled for this reason: Testing.\\" iconSize={48} layout=\\"vertical\\">
                              <div className=\\"bp4-non-ideal-state bp4-non-ideal-state-vertical\\">
                                <div className=\\"bp4-non-ideal-state-visual\\" style={{...}}>
                                  <Blueprint4.Icon icon=\\"error\\" size={48} aria-hidden={true} tabIndex={-1}>
                                    <span aria-hidden={true} icon=\\"error\\" tabIndex={-1} className=\\"bp4-icon bp4-icon-error\\" title={[undefined]}>
                                      <svg fill={[undefined]} data-icon=\\"error\\" width={48} height={48} viewBox=\\"0 0 20 20\\" aria-labelledby={[undefined]} role=\\"img\\">
                                        <path d=\\"M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z\\" fillRule=\\"evenodd\\" />
                                      </svg>
                                    </span>
                                  </Blueprint4.Icon>
                                </div>
                                <div className=\\"bp4-non-ideal-state-text\\">
                                  <Component>
                                    <h4 className=\\"bp4-heading\\">
                                      Disabled
                                    </h4>
                                  </Component>
                                  <div>
                                    The Source Academy has been disabled for this reason: Testing.
                                  </div>
                                </div>
                              </div>
                            </Blueprint4.NonIdealState>
                          </div>
                        </Disabled>
                      </RenderedRoute>
                    </Outlet>
                  </div>
                </div>
              </Application>
            </RenderedRoute>
          </RenderErrorBoundary>
        </DataRoutes>
      </Router>
    </RouterProvider>
  </ApplicationWrapper>
</Provider>"
`;
